data-exporter/src/main/java/org/reactome/release/dataexport/UniProtReactomeEntry.java

```diff
@@ -205,46 +208,143 @@ private UniProtReactomeEntry(long dbId, String accession, String displayName) {
+	 * representing all associated events. The events will include both ReactionlikeEvents, as well as Pathways
+	 * containing the ReactionlikeEvents, which correspond to the passed ReactionlikeEvent db ids.
+	 * @param reactionLikeEventIds ReactionlikeEvent db ids for which to retrieve associated ReactomeEvent objects
+	 * @param graphDBSession Neo4J Driver Session object for querying the graph database
+	 * @return Set of ReactomeEvents representing the ReactionlikeEvents and Pathways containing the ReactionlikeEvents
+	 * which correspond to the passed ReactionlikeEvent db ids
+	 */
+	private static Set<ReactomeEvent> getRLEAndPathwayReactomeEventsFromRLEIds(
+		Set<Long> reactionLikeEventIds, Session graphDBSession
+	) {
+		Set<ReactomeEvent> reactomeEvents = new HashSet<>();
+
+		reactomeEvents.addAll(getRLEReactomeEventsFromRLEIds(reactionLikeEventIds, graphDBSession));
+		reactomeEvents.addAll(getPathwayReactomeEventsFromRLEIds(reactionLikeEventIds, graphDBSession));
+
+		return reactomeEvents;
+	}
+
+	/**
+	 * Retrieves, from the graph database for the passed ReactionlikeEvent db ids, the set of ReactomeEvent objects
+	 * which represent the ReactionlikeEvents.
+	 * @param reactionLikeEventIds ReactionlikeEvent db ids for which to retrieve associated ReactomeEvent objects
+	 * @param graphDBSession Neo4J Driver Session object for querying the graph database
+	 * @return Set of ReactomeEvents representing the ReactionlikeEvents which correspond to the passed
+	 * ReactionlikeEvent db ids
+	 */
+	private static Set<ReactomeEvent> getRLEReactomeEventsFromRLEIds(
+		Set<Long> reactionLikeEventIds, Session graphDBSession
+	) {
+		return new HashSet<>(convertDbIdsToReactomeEvents(reactionLikeEventIds, graphDBSession));
+	}
+
+	/**
+	 * Retrieves, from the graph database for the passed ReactionlikeEvent db ids, the set of ReactomeEvent objects
+	 * represent Reactome Pathways containing the ReactionlikeEvents
+	 * @param reactionLikeEventIds ReactionlikeEvent db ids for which to retrieve associated ReactomeEvent objects
+	 * representing pathways
+	 * @param graphDBSession Neo4J Driver Session object for querying the graph database
+	 * @return Set of ReactomeEvents representing the Pathways containing the ReactionlikeEvents
+	 * which correspond to the passed ReactionlikeEvent db ids
+	 */
+	private static Set<ReactomeEvent> getPathwayReactomeEventsFromRLEIds(
+		Set<Long> reactionLikeEventIds, Session graphDBSession
+	) {
+		Set<ReactomeEvent> pathwayReactomeEvents = new HashSet<>();
+
+		pathwayReactomeEvents.addAll(reactionLikeEventIds.stream().flatMap(
+			rleId -> {
+				Map<Long, Set<Long>> rleIdToPathwayIds = fetchRLEIdToPathwayId(graphDBSession);
+				Set<Long> pathwayIds = rleIdToPathwayIds.computeIfAbsent(rleId, k -> new HashSet<>());
+				return pathwayIds.stream();
+			}
+		).map(pathwayId -> convertDbIdToReactomeEvent(pathwayId, graphDBSession)).collect(Collectors.toList()));
```

Comment:
So, you map Event IDs to... Pathway IDs. Then you map those to... Events? Could this get a little more documentation?

Also: why two separate calls to flatMap/map? Why not just have one mapping function?

Reply:
Okay, I reworked it so it should be clearer now.